message/validation/validator.go

```diff
@@ -0,0 +1,467 @@
+		return fmt.Errorf("size exceeded")
+	}
+
+	if len(signedMsg.Signature) != 96 {
+		return fmt.Errorf("wrong signature size: %d", len(signedMsg.Signature))
+	}
+
+	// TODO: check if has running duty
+
+	msgID := msg.GetID()
+
+	// Validate that the specified signers belong in the validator's committee.
+	if err := mv.validSigners(share, msg); err != nil {
+		return err
+	}
+
+	slot := phase0.Slot(signedMsg.Message.Height)
+
+	if mv.earlyMessage(slot) {
+		return ErrEarlyMessage
+	}
+	if mv.lateMessage(slot, msgID.GetRoleType()) {
+		return ErrLateMessage
+	}
+
+	if bytes.Equal(signedMsg.Signature, bytes.Repeat([]byte{0}, len(signedMsg.Signature))) {
+		return fmt.Errorf("zero signature")
+	}
+
+	pj, err := signedMsg.Message.GetPrepareJustifications()
+	if err != nil {
+		return fmt.Errorf("malfrormed prepare justifications: %w", err)
+	}
+
+	rcj, err := signedMsg.Message.GetRoundChangeJustifications()
+	if err != nil {
+		return fmt.Errorf("malfrormed round change justifications: %w", err)
+	}
+
+	if err := ssvtypes.VerifyByOperators(signedMsg.Signature, signedMsg, mv.netCfg.Domain, spectypes.QBFTSignatureType, share.Committee); err != nil {
+		return fmt.Errorf("invalid signature: %w", err)
+	}
+
+	// TODO: checks for pj and rcj
+	_ = pj
+	_ = rcj
+
+	consensusID := ConsensusID{
+		PubKey: phase0.BLSPubKey(msgID.GetPubKey()),
+		Role:   msgID.GetRoleType(),
+	}
+	// Validate each signer's behavior.
+	state := mv.consensusState(consensusID)
+	for _, signer := range signedMsg.Signers {
+		if err := mv.validateSignerBehavior(state, signer, msg); err != nil {
+			return fmt.Errorf("bad signed behavior: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (mv *MessageValidator) validatePartialSignatureMessage(msg *queue.DecodedSSVMessage) error {
+	signedMsg, ok := msg.Body.(*spectypes.PartialSignatureMessage)
+	if !ok {
+		return fmt.Errorf("expected partial signature message")
+	}
+
+	if len(msg.Data) > maxPartialSignatureMsgSize {
+		return fmt.Errorf("size exceeded") // ERR_PSIG_MSG_SIZE
+	}
+	_ = signedMsg // TODO: validate it
+
+	// TODO: check signature
+
+	return nil
+}
+func (mv *MessageValidator) validateEventMessage(msg *queue.DecodedSSVMessage) error {
+	signedMsg, ok := msg.Body.(*ssvtypes.EventMsg)
+	if !ok {
+		return fmt.Errorf("expected event message")
+	}
+
+	_ = signedMsg // TODO: validate it
+
+	return nil
+}
+
+func (mv *MessageValidator) earlyMessage(slot phase0.Slot) bool {
+	return mv.netCfg.Beacon.GetSlotEndTime(mv.netCfg.Beacon.EstimatedCurrentSlot()).
+		Add(-clockErrorTolerance).Before(mv.netCfg.Beacon.GetSlotStartTime(slot))
+}
+
+func (mv *MessageValidator) lateMessage(slot phase0.Slot, role spectypes.BeaconRole) bool {
+	// Note: this is just an example, should be according to ssv-spec.
+	var ttl phase0.Slot
+	switch role {
+	case spectypes.BNRoleProposer, spectypes.BNRoleSyncCommittee, spectypes.BNRoleSyncCommitteeContribution:
+		ttl = 1 + 1
+	case spectypes.BNRoleAttester, spectypes.BNRoleAggregator:
+		ttl = 32 + 2
```

Comment:
Both should be + 2 (so 1 + 2 and 32 + 2). My spec was confusing and wrong here, I'll update there. Sorry for that.
The +2 is a common slot spread that Prysm suggested, so it can even be a const variable (config, global, as you want and if you want).